### PR TITLE
Enable carrier waiting after interface up

### PR DESCRIPTION
**Pull request by xenolog**
https://blueprints.launchpad.net/fuel/+spec/l23network-carrier-check-after-ifup

### CODE REVIEW

File: deployment/puppet/l23network/lib/puppet/provider/l3_if_downup/ruby.rb [hunk 1]
```diff
@@ -121,11 +122,43 @@ def restart()
...
+            loop do
+              begin
+                _rc = File.open("/sys/class/net/#{@resource[:interface]}/carrier", 'r'){ |file| file.read() }
+              rescue
+                _rc = -1
+              end
+              if _rc.to_i() == 1
+                break
+              elsif _rc.to_i() == -1
+                notice("Interface '#{@resource[:interface]}' prohibited by administrator!!!")
```
Comment (xdel) [inline deployment/puppet/l23network/lib/puppet/provider/l3_if_downup/ruby.rb]: Please change remark to 'Seems that the interface was brought down administratively. Further deployment actions may fail'.

File: deployment/puppet/l23network/lib/puppet/type/l3_if_downup.rb [hunk 2]
```diff
@@ -57,8 +57,29 @@
...
+    end
+
+    newparam(:wait_carrier_after_ifup_timeout) do
+      desc "Timeout for carrier waiting after interface up."
+      defaultto(120)
+      validate do |val|
+        if val.to_i() >= 0
+          true
+        else
+          fail("Timeout must be in seconds, not '#{val}'")
```
Comment (xdel) [inline deployment/puppet/l23network/lib/puppet/type/l3_if_downup.rb]: must be positive integer

File: deployment/puppet/l23network/lib/puppet/provider/l3_if_downup/ruby.rb [hunk 3]
```diff
@@ -121,11 +122,43 @@ def restart()
...
       end
       notice("Interface '#{@resource[:interface]}' up.")
+      # checking and waiting carrier for PHYS. interface
+      if (@resource[:interface] =~ /^eth\d+$/) and @resource[:wait_carrier_after_ifup]
+        begin
+          Timeout::timeout(@resource[:wait_carrier_after_ifup_timeout]) do
+            _w = 10
+            loop do
+              begin
+                _rc = File.open("/sys/class/net/#{@resource[:interface]}/carrier", 'r'){ |file| file.read() }
```
Comment (xdel) [inline deployment/puppet/l23network/lib/puppet/provider/l3_if_downup/ruby.rb:137]: I`d throw a readable error if file is not available.

Comment (xenolog) [inline deployment/puppet/l23network/lib/puppet/provider/l3_if_downup/ruby.rb:137]: see about 10 lines bellow.

File: deployment/puppet/l23network/lib/puppet/provider/l3_if_downup/ruby.rb (hunk 1 above)
Comment (xenolog) [inline deployment/puppet/l23network/lib/puppet/provider/l3_if_downup/ruby.rb]: fixed

File: deployment/puppet/l23network/lib/puppet/type/l3_if_downup.rb (hunk 2 above)
Comment (xenolog) [inline deployment/puppet/l23network/lib/puppet/type/l3_if_downup.rb]: fixed

**Comment by aglarendil** [comment]
+2

**Pull request merged**